base/strings/unicode.jl

```diff
@@ -150,22 +150,45 @@ end
 
 utf8proc_map(s::AbstractString, flags::Integer) = utf8proc_map(String(s), flags)
 
-# Documented in Unicode module
-function normalize(
-    s::AbstractString;
-    stable::Bool=false,
-    compat::Bool=false,
-    compose::Bool=true,
-    decompose::Bool=false,
-    stripignore::Bool=false,
-    rejectna::Bool=false,
-    newline2ls::Bool=false,
-    newline2ps::Bool=false,
-    newline2lf::Bool=false,
-    stripcc::Bool=false,
-    casefold::Bool=false,
-    lump::Bool=false,
-    stripmark::Bool=false,
+
+function utf8proc_map(str::String, options::Integer, custom_func, custom_data)
```

Comment:
On the Julia side you should just pass `custom_func` — there is no need for `custom_data` in Julia because that is just a crude way to simulate a closure in C.

Then write
```
custom_func_wrapper(codepoint::Int32, custom_func::Any) = Int32(custom_func(codepoint))
```
and in `utf8proc_map` do
```
ccustom_func = @cfunction(custom_func_wrapper, Int32, (Int32, Any))
```
and pass `ccustom_func, custom_func` to `utf8proc_decompose_custom`.   (That is, the Julia `custom_func` is actually passed as the *data* to `utf8proc_decompose_custom`.   The corresponding argument in the `ccall` should be defined as `Any`.)

Reply:
Oh, That is surprising. If it is the case, do we really need to make `custom_func` a closure?
But beside the origin of the `custom_data`,  will it be more convenient to pass extra data for the `custom_func`? I read the c code at [utf8proc](https://github.com/JuliaStrings/utf8proc/). It seems that the `custom_func` can only do a codepoint mapping? I original thought that the `custom_data` is something like a pre-define table for mapping the codepoint and can change the behavior of `utf8proc_iterate`.

Reply:
You don’t need an extra argument to pass extra data to a function argument in Julia. You just pass an anonymous function that references whatever local data you want. 

In general, these “data” arguments are only for languages like C that don’t have closures.  Google “closure computer science” if you don’t know what I’m talking about.

Reply:
I just wondering whether using closure is a good design. Suppose we have a struct called `MappingRule` and a method `mapping(m::MappingRule, codepoing)`. Then every time you call the `normalize` you will need to pass a `x->mapping(m, x)`?

Reply:
Yes.   Or you just make the `MappingRule` a callable object.